### ISSUE TITLE
Type problems with using __init__ / class constructor

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1196]
It seems that Pylance might have a bug when inferring the type of a class initializer (`__init__`). In particular when being fed iterable input returned by builtins such as `zip` and `enumerate`. 

## Environment data

Pylance language server 2020.11.2 (pyright 750a296b)
OS and version: Mac OS 11.01
Python version (& distribution if applicable, e.g. Anaconda): Python 3.9 (brew)

## Expected behaviour

Expect Pylance to infer the right type of class initializer:

<img width="630" alt="Screenshot 2020-11-21 at 17 34 33" src="https://user-images.githubusercontent.com/849479/99882343-f9b61480-2c1f-11eb-9091-c00d88aca735.png">

## Actual behaviour

<img width="1195" alt="Screenshot 2020-11-21 at 17 34 42" src="https://user-images.githubusercontent.com/849479/99882345-03d81300-2c20-11eb-97c9-d9c0e6ce4fff.png">

One can see where that the inferred type is very wrong i.e `Type "Tuple[int, Tuple[int, Tuple[int, TSource]]]`. 

A mitigation I have found is to call the constructor through a static create method that specifies the return type correctly:

<img width="597" alt="Screenshot 2020-11-21 at 17 35 01" src="https://user-images.githubusercontent.com/849479/99882350-105c6b80-2c20-11eb-8a64-7e4c359e6c37.png">

## Logs

Nothing unusual:

```
[Info  - 5:17:27 PM] Pylance language server 2020.11.2 (pyright 750a296b) starting
[Info  - 5:17:27 PM] Server root directory: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist
[Error - 5:17:27 PM] stubPath /Users/dbrattli/Developer/Github/aioreactive/typings is not a valid directory.
[Info  - 5:41:10 PM] No configuration file found.
[Info  - 5:41:10 PM] Setting pythonPath for service "aioreactive": "/usr/local/opt/python@3.9/bin/python3"
Search paths found for configured python interpreter:
  /usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9
  /usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload
  /Users/dbrattli/Library/Python/3.9/lib/python/site-packages
  /usr/local/lib/python3.9/site-packages
  /Users/dbrattli/Developer/Github/aioreactive
  /Users/dbrattli/Developer/Github/Expression
  /Users/dbrattli/Developer/Github/RxPY
  /usr/local/Cellar/protobuf/3.13.0_1/libexec/lib/python3.9/site-packages
[Error - 5:41:10 PM] stubPath /Users/dbrattli/Developer/Github/aioreactive/typings is not a valid directory.
[Info  - 5:41:10 PM] Assuming Python version 3.9
[Info  - 5:41:10 PM] Assuming Python platform Darwin
[Info  - 5:41:10 PM] Searching for source files
[Info  - 5:41:10 PM] Found 53 source files
Background analysis message: setConfigOptions
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: analyze
[BG(1)] analyzing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py ...
[BG(1)]   parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py (40ms)
[BG(1)]   parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/builtins.pyi [fs read 1ms] (37ms)
[BG(1)]   binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/builtins.pyi (10ms)
[BG(1)]   binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py (4ms)
[BG(1)]   checking: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py ...
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/typing.pyi [fs read 1ms] (11ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/typing.pyi (3ms)
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi [fs read 1ms] (9ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (1ms)
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/collections/__init__.py [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/collections/__init__.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/collections/seq.py [fs read 1ms] (13ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/collections/seq.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/__init__.py [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/__init__.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/mailbox.py [fs read 1ms] (10ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/mailbox.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/option.py [fs read 1ms] (5ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/option.py (2ms)
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/abc.pyi [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3/abc.pyi (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/match.py [fs read 0ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/match.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/error.py [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/error.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/result.py [fs read 0ms] (5ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/result.py (2ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/fn.py [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/fn.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/compose.py [fs read 1ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/compose.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/core/pipe.py [fs read 1ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/core/pipe.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/system/__init__.py [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/system/__init__.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/system/disposable.py [fs read 1ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/system/disposable.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/combine.py [fs read 0ms] (8ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/combine.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/types.py [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/types.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/2and3/logging/__init__.pyi [fs read 1ms] (8ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/2and3/logging/__init__.pyi (2ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/create.py [fs read 0ms] (4ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/create.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/msg.py [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/msg.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3.7/dataclasses.pyi [fs read 1ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.11.2/dist/typeshed-fallback/stdlib/3.7/dataclasses.pyi (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/observables.py [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/observables.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/observers.py [fs read 0ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/observers.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/utils.py [fs read 0ms] (1ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/utils.py (0ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/notification.py [fs read 1ms] (6ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/notification.py (1ms)
[BG(1)]     parsing: /Users/dbrattli/Developer/Github/Expression/expression/system/cancellation.py [fs read 1ms] (2ms)
[BG(1)]     binding: /Users/dbrattli/Developer/Github/Expression/expression/system/cancellation.py (1ms)
[BG(1)]   checking: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py (349ms)
[BG(1)] analyzing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/transform.py (441ms)
Background analysis message: resumeAnalysis
[BG(1)] analyzing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/testing/test.py ...
[BG(1)]   parsing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/testing/test.py (0ms)
[BG(1)]   binding: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/testing/test.py (0ms)
[BG(1)]   checking: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/testing/test.py (27ms)
[BG(1)] analyzing: /Users/dbrattli/Developer/Github/aioreactive/aioreactive/testing/test.py (27ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.89, which I just published. It will also appear in the next Pylance release.
